shanghai/core_plugins/channel.py

```diff
@@ -0,0 +1,305 @@
+# Copyright © 2016  Lars Peter Søndergaard <lps@chireiden.net>
+# Copyright © 2016  FichteFoll <fichtefoll2@googlemail.com>
+#
+# This file is part of Shanghai, an asynchronous multi-server IRC bot.
+#
+# Shanghai is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Shanghai is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Shanghai.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import string
+
+from shanghai.event import global_event, GlobalEventName
+from shanghai.network import NetworkContext
+from shanghai.irc.server_reply import ServerReply
+from .message import Message
+
+__plugin_name__ = 'Channel'
+__plugin_version__ = '0.1.0'
+__plugin_description__ = 'Track channel state'
+
+
+class _Base:
+    def __hash__(self):
+        return hash(id(self))
+
+
+class Channel(_Base):
+    def __init__(self, name, modes=None):
+        self.name = name
+        self.modes = modes
+
+
+class User(_Base):
```

Comment:
Isn't this exactly what `shangahi.irc.Prefix` is for? They are identical, except that Prefix's `__init__` has no defaults, which can be added easily.